flink-core/src/main/java/org/apache/flink/configuration/MemorySize.java

```diff
@@ -117,7 +121,30 @@ public boolean equals(Object obj) {
 
 	@Override
 	public String toString() {
-		return bytes + " bytes";
+		List<MemoryUnit> orderedUnits = Arrays.asList(
+			BYTES,
+			KILO_BYTES,
+			MEGA_BYTES,
+			GIGA_BYTES,
+			TERA_BYTES);
+
+		MemoryUnit highestIntegerUnit = IntStream.range(0, orderedUnits.size())
+			.sequential()
+			.filter(idx -> bytes % orderedUnits.get(idx).getMultiplier() != 0)
+			.boxed()
+			.findFirst()
+			.map(idx -> {
+				if (idx == 0) {
+					return orderedUnits.get(0);
+				} else {
+					return orderedUnits.get(idx - 1);
+				}
+			}).orElse(BYTES);
+
+		return String.format(
+			"%d %s",
+			bytes / highestIntegerUnit.getMultiplier(),
+			highestIntegerUnit.getUnits()[1]);
```

Comment:
Since this is a rather complex operation, I would suggest to cache the result similar to what we are doing in `AbstractID`.